### ISSUE TITLE
Principle mistake in blog code example

### ISSUE THREAD

**Issue by lve0200** [rstudio/tensorflow-blog#30]
In your article
Chollet & Allaire (2017, Dec. 20). TensorFlow for R: Time Series Forecasting with Recurrent Neural Networks. Retrieved from https://blogs.rstudio.com/tensorflow/posts/2017-12-20-time-series-forecasting-with-recurrent-neural-networks/
the code of the generator function appears to be untested! The generator function works on features in columns and samples in rows, but the data shown at the top and as explained/treated in the text is the other way around, samples in col and features in rows. The generator tries to access rows as samples, since there are only 16 Features, it immediately runs out of boundary
RGDS
hk

**Comment by skeydan** [comment]
I've updated the generator code as per the book's errata: https://manning-content.s3.amazonaws.com/download/1/6d1e9c3-9483-455c-8a47-54bb4bbd3205/Allaire_DeepLearningwithR_err4.html.
See https://github.com/rstudio/tensorflow-blog/pull/31.

Otherwise, can you describe what you are referring to exactly? Which code blocks/graphics/text exactly do you mean?

**Comment by lve0200** [comment]
WRT the a.m. article. Unfortunately the code segments and pictures are not labelled, hence I cannot reference these. Therefore I copy them here as code. The following table shows data organised as following:
Features (Date_, d, T, Tpot...) are in ROWs and there are 420,551 Observations of these in  columns. The value to be predicted by the NN is the Temperature (T), i.e. the 3rd row.

> 

`data_dir <- "~/Downloads/jena_climate"
fname <- file.path(data_dir, "jena_climate_2009_2016.csv")
data <- read_csv(fname)

glimpse(data)

Observations: 420,551
Variables: 15
$ `Date Time`       <chr> "01.01.2009 00:10:00", "01.01.2009 00:20:00", "...
$ `p (mbar)`        <dbl> 996.52, 996.57, 996.53, 996.51, 996.51, 996.50,...
$ `T (degC)`        <dbl> -8.02, -8.41, -8.51, -8.31, -8.27, -8.05, -7.62...
$ `Tpot (K)`        <dbl> 265.40, 265.01, 264.91, 265.12, 265.15, 265.38,...
$ `Tdew (degC)`     <dbl> -8.90, -9.28, -9.31, -9.07, -9.04, -8.78, -8.30...
$ `rh (%)`          <dbl> 93.3, 93.4, 93.9, 94.2, 94.1, 94.4, 94.8, 94.4,...
$ `VPmax (mbar)`    <dbl> 3.33, 3.23, 3.21, 3.26, 3.27, 3.33, 3.44, 3.44,...
$ `VPact (mbar)`    <dbl> 3.11, 3.02, 3.01, 3.07, 3.08, 3.14, 3.26, 3.25,...
$ `VPdef (mbar)`    <dbl> 0.22, 0.21, 0.20, 0.19, 0.19, 0.19, 0.18, 0.19,...
$ `sh (g/kg)`       <dbl> 1.94, 1.89, 1.88, 1.92, 1.92, 1.96, 2.04, 2.03,...
$ `H2OC (mmol/mol)` <dbl> 3.12, 3.03, 3.02, 3.08, 3.09, 3.15, 3.27, 3.26,...
$ `rho (g/m**3)`    <dbl> 1307.75, 1309.80, 1310.24, 1309.19, 1309.00, 13...
$ `wv (m/s)`        <dbl> 1.03, 0.72, 0.19, 0.34, 0.32, 0.21, 0.18, 0.19,...
$ `max. wv (m/s)`   <dbl> 1.75, 1.50, 0.63, 0.50, 0.63, 0.63, 0.63, 0.50,...
$ `wd (deg)`        <dbl> 152.3, 136.1, 171.6, 198.0, 214.3, 192.7, 166.5...`

<

The generator is supposed to extract chunks of data, 2b fed into the NN. However, the way the gen appears to work is, that it composes these batches of data in the wrong sense. The generator function works on features(Variables) in columns and samples(Observations) in rows, but the data shown at the top and as explained/treated in the text of the article is the other way around, samples in col and features in rows.
Moreover, the samples generated by the generator contains the input variables as well as the target Temperature values. Hence the targets are duplicated in targets and again in samples.

> 
      samples[j,,] <- data[indices,]
      targets[[j]] <- data[rows[[j]] + delay,2]
<
Please understand that the plain vanilla code of the generator may be syntactically OK, but the way it is applied onto the existing data appears 2b wrong.
RGDS
hk

**Comment by skeydan** [comment]
Now I see what you mean. 

This 

```
Observations: 420,551
Variables: 15
$ `Date Time`       <chr> "01.01.2009 00:10:00", "01.01.2009 00:20:00", "...
$ `p (mbar)`        <dbl> 996.52, 996.57, 996.53, 996.51, 996.51, 996.50,...
$ `T (degC)`        <dbl> -8.02, -8.41, -8.51, -8.31, -8.27, -8.05, -7.62...
```

is just the way `glimpse` displays a concise ... well ... glimpse of the data. I must say I use it a lot to get a quick impression of the data.

Reg. 

> Hence the targets are duplicated in targets and again in samples.

this will depend on what you actually intend to include among the predictors. Here we definitely want to include previous temperatures.

**Comment by lve0200** [comment]
Ah, OK,
may you should add a hint that the original data is different from what glimpse displays. There may be other people(beginners like me), who work on a different data set and start wondering about col and rows!

**Comment by lve0200** [comment]
Sorry, but 2 more issues, probably due to my ignorance:

> train_data <- data[1:200000,]
mean <- apply(train_data, 2, mean)
std <- apply(train_data, 2, sd)
data <- scale(data, center = mean, scale = std)
> 
You calc mean/std for the training subset "train_data", but then you scale the entire data frame "data" with it. Is this intended? Why?
For generating the sets, you have Shuffle=True for training (but not for the others); doesn't this break the sequentiality, we're after?

> train_gen <- generator(
  data,
  ...
  shuffle = TRUE,
  step = step, 
  batch_size = batch_size
)

val_gen = generator(
  data,
  lookback = lookback,
  delay = delay,
  min_index = 200001,
  max_index = 300000,
  step = step,
  batch_size = batch_size
)

test_gen <- generator(
  data,
  lookback = lookback,
  delay = delay,
  min_index = 300001,
  max_index = NULL,
  step = step,
  batch_size = batch_size
)

**Comment by skeydan** [comment]
> You calc mean/std for the training subset "train_data", but then you scale the entire data frame "data" with it. Is this intended? Why?

This is common practice in ML; you compute statistics on the training data and then use that to normalize the test set too.

> For generating the sets, you have Shuffle=True for training (but not for the others); doesn't this break the sequentiality, we're after?

The LSTM works with sequences of a definable length; then the sequences themselves are shuffled.

**Comment by lve0200** [comment]
OK, vielen Dank, nochmal und Servus nach München

**Comment by skeydan** [comment]
ciao, ciao
